toolkit/tools/pkg/imagecustomizerlib/liveosisobuilder.go

```diff
@@ -0,0 +1,671 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/configuration"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safemount"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/pkg/isomakerlib"
+)
+
+var (
+	grubCfgTemplate = `set default="0"
+set timeout=0
+
+menuentry "Mariner Baremetal Iso" {
+
+	search --label CDROM --set root
+	linux /isolinux/vmlinuz \
+			overlay-size=70% \
+			selinux=0 \
+			console=tty0 \
+			apparmor=0 \
+			root=live:LABEL=CDROM \
+			rd.shell \
+			rd.live.image \
+			rd.live.dir=config/additionalfiles/0 \
+			rd.live.squashimg=rootfs.img \
+			rd.live.overlay=1 \
+			rd.live.overlay.nouserconfirmprompt
+
+	initrd /isolinux/initrd.img
+}	
+`
+	dracutConfig = `add_dracutmodules+=" dmsquash-live "
+add_drivers+=" overlay "
+`
+)
+
+type IsoWorkingDirs struct {
+	isoBuildDir string
+	// 'isomakerBuildDir' will be deleted/re-created by IsoMaker before it
+	// proceeds. It needs to be different from `isoBuildDir`.
+	isomakerBuildDir string
+	outDir           string
+}
+
+// `IsoArtifacts` holds the extracted/generated artifacts necessary to build
+// a LiveOS ISO image.
+type IsoArtifacts struct {
+	kernelVersion     string
+	bootx64EfiPath    string
+	grubx64EfiPath    string
+	grubCfgPath       string
+	vmlinuzPath       string
+	initrdImagePath   string
+	squashfsImagePath string
+}
+
+type LiveOSIsoBuilder struct {
+	workingDirs IsoWorkingDirs
+	artifacts   IsoArtifacts
+}
+
+// extractArtifactsFromBootDevice
+//
+//	extracts the bootloaders from the specified boot device.
+//
+// inputs:
+//   - 'bootDevicePath': path to an existing boot device.
+//   - 'bootfsType': file system type of the specified boot device.
+//
+// output:
+//
+//	the bootloaders are saved to the b.workingDirs.isoBuildDir
+func (b *LiveOSIsoBuilder) extractArtifactsFromBootDevice(bootDevicePath string, bootfsType string) error {
+
+	logger.Log.Infof("Extracting artifacts from the boot partition...")
+
+	loopDevMountFullDir := filepath.Join(b.workingDirs.isoBuildDir, "readonly-boot-mount")
+	logger.Log.Infof("Mounting %s(%s) to %s", bootDevicePath, bootfsType, loopDevMountFullDir)
+
+	fullDiskBootMount, err := safemount.NewMount(bootDevicePath, loopDevMountFullDir, bootfsType, 0, "", true)
```

Comment:
hmm... it's causing subsequent mounting to fail saying the device is read-only. Maybe it's already mounted r/w. I'll look more into this.

Reply:
You should avoid mounting something more than once concurrently, as it can cause data corruption. (You can use a bind mount if you need to add the mount to another directory.)

Reply:
I've found out that `connectToExistingImage`  mounts all partitions before it returns. So, I've removed my mounting code entirely.